In [74]:
import pandas as pd
import numpy as np
from scipy import stats
import matplotlib.pyplot as plt
import seaborn as sns
import statsmodels.api as sm
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.ensemble import GradientBoostingRegressor
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.ensemble import RandomForestRegressor
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import AdaBoostRegressor
from sklearn.ensemble import AdaBoostClassifier
from sklearn import metrics
from sklearn.metrics import mean_squared_error
from xgboost.sklearn import XGBRegressor

In [75]:
sales=pd.read_csv("C:/Users/Jothy Natarajan/Downloads/bank.csv",sep=";")

In [76]:
sales.shape

(41188, 21)

In [77]:
sales.head(20)

age          job   marital            education  default housing loan  \
0    56    housemaid   married             basic.4y       no      no   no   
1    57     services   married          high.school  unknown      no   no   
2    37     services   married          high.school       no     yes   no   
3    40       admin.   married             basic.6y       no      no   no   
4    56     services   married          high.school       no      no  yes   
5    45     services   married             basic.9y  unknown      no   no   
6    59       admin.   married  professional.course       no      no   no   
7    41  blue-collar   married              unknown  unknown      no   no   
8    24   technician    single  professional.course       no     yes   no   
9    25     services    single          high.school       no     yes   no   
10   41  blue-collar   married              unknown  unknown      no   no   
11   25     services    single          high.school       no     yes   no   
12   29  blue-collar    single          high.school       no      no  yes   
13   57    housemaid  divorced             basic.4y       no     yes   no   
14   35  blue-collar   married             basic.6y       no     yes   no   
15   54      retired   married             basic.9y  unknown     yes  yes   
16   35  blue-collar   married             basic.6y       no     yes   no   
17   46  blue-collar   married             basic.6y  unknown     yes  yes   
18   50  blue-collar   married             basic.9y       no     yes  yes   
19   39   management    single             basic.9y  unknown      no   no   

      contact month day_of_week  ...  campaign  pdays  previous     poutcome  \
0   telephone   may         mon  ...         1    999         0  nonexistent   
1   telephone   may         mon  ...         1    999         0  nonexistent   
2   telephone   may         mon  ...         1    999         0  nonexistent   
3   telephone   may         mon  ...         1    999         0  nonexistent   
4   telephone   may         mon  ...         1    999         0  nonexistent   
5   telephone   may         mon  ...         1    999         0  nonexistent   
6   telephone   may         mon  ...         1    999         0  nonexistent   
7   telephone   may         mon  ...         1    999         0  nonexistent   
8   telephone   may         mon  ...         1    999         0  nonexistent   
9   telephone   may         mon  ...         1    999         0  nonexistent   
10  telephone   may         mon  ...         1    999         0  nonexistent   
11  telephone   may         mon  ...         1    999         0  nonexistent   
12  telephone   may         mon  ...         1    999         0  nonexistent   
13  telephone   may         mon  ...         1    999         0  nonexistent   
14  telephone   may         mon  ...         1    999         0  nonexistent   
15  telephone   may         mon  ...         1    999         0  nonexistent   
16  telephone   may         mon  ...         1    999         0  nonexistent   
17  telephone   may         mon  ...         1    999         0  nonexistent   
18  telephone   may         mon  ...         1    999         0  nonexistent   
19  telephone   may         mon  ...         1    999         0  nonexistent   

   emp.var.rate  cons.price.idx  cons.conf.idx  euribor3m  nr.employed   y  
0           1.1          93.994          -36.4      4.857       5191.0  no  
1           1.1          93.994          -36.4      4.857       5191.0  no  
2           1.1          93.994          -36.4      4.857       5191.0  no  
3           1.1          93.994          -36.4      4.857       5191.0  no  
4           1.1          93.994          -36.4      4.857       5191.0  no  
5           1.1          93.994          -36.4      4.857       5191.0  no  
6           1.1          93.994          -36.4      4.857       5191.0  no  
7           1.1          93.994          -36.4      4.857       5191.0  no  
8           

In [26]:
sales.rename(columns={'emp.var.rate':'emp_var_rate','cons.price.idx':'cons_price_idx','cons.conf.idx':'cons_conf_idx','nr.employed':'nr_employed'},inplace=True)

In [27]:
sales.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 41188 entries, 0 to 41187
Data columns (total 21 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   age             41188 non-null  int64  
 1   job             41188 non-null  int32  
 2   marital         41188 non-null  int32  
 3   education       41188 non-null  int32  
 4   default         41188 non-null  int32  
 5   housing         41188 non-null  int32  
 6   loan            41188 non-null  int32  
 7   contact         41188 non-null  int32  
 8   month           41188 non-null  int32  
 9   day_of_week     41188 non-null  int32  
 10  duration        41188 non-null  int64  
 11  campaign        41188 non-null  int64  
 12  pdays           41188 non-null  int64  
 13  previous        41188 non-null  int64  
 14  poutcome        41188 non-null  int32  
 15  emp_var_rate    41188 non-null  float64
 16  cons_price_idx  41188 non-null  float64
 17  cons_conf_idx   41188 non-null 

<H2>Q1. HYPOTHESIS TESTING </H2>

In [80]:
yes1=sales.y[sales.default=='yes']
no1=sales.y[sales.default=='no']
unknown1=sales.y[sales.default=='unknown']

In [82]:
from statsmodels.stats.proportion import proportions_ztest
table=pd.crosstab(yes1,no1,unknown1)
stats.chi2_contingency(table)

ValueError: values cannot be used without an aggfunc.

<h3>2) a)Missing Value Analysis</h3>

In [28]:
sales.isnull().sum()

age               0
job               0
marital           0
education         0
default           0
housing           0
loan              0
contact           0
month             0
day_of_week       0
duration          0
campaign          0
pdays             0
previous          0
poutcome          0
emp_var_rate      0
cons_price_idx    0
cons_conf_idx     0
euribor3m         0
nr_employed       0
y                 0
dtype: int64

<h3>There is no null values</h3>

<h3>2) b. Label Encoding  </h3>

In [17]:
from sklearn import preprocessing
label_encoder = preprocessing.LabelEncoder() 
sales['job']= label_encoder.fit_transform(sales['job'])
sales['marital']= label_encoder.fit_transform(sales['marital'])
sales['education']= label_encoder.fit_transform(sales['education'])
sales['default']= label_encoder.fit_transform(sales['default'])
sales['housing']= label_encoder.fit_transform(sales['housing'])
sales['loan']= label_encoder.fit_transform(sales['loan'])
sales['contact']= label_encoder.fit_transform(sales['contact'])
sales['month']= label_encoder.fit_transform(sales['month'])
sales['day_of_week']= label_encoder.fit_transform(sales['day_of_week'])
sales['poutcome']= label_encoder.fit_transform(sales['poutcome'])

In [18]:
sales.head(5)

age  job  marital  education  default  housing  loan  contact  month  \
0   56    3        1          0        0        0     0        1      6   
1   57    7        1          3        1        0     0        1      6   
2   37    7        1          3        0        2     0        1      6   
3   40    0        1          1        0        0     0        1      6   
4   56    7        1          3        0        0     2        1      6   

   day_of_week  ...  campaign  pdays  previous  poutcome  emp.var.rate  \
0            1  ...         1    999         0         1           1.1   
1            1  ...         1    999         0         1           1.1   
2            1  ...         1    999         0         1           1.1   
3            1  ...         1    999         0         1           1.1   
4            1  ...         1    999         0         1           1.1   

   cons.price.idx  cons.conf.idx  euribor3m  nr.employed   y  
0          93.994          -36.4      4.857       5191.0  no  
1          93.994          -36.4      4.857       5191.0  no  
2          93.994          -36.4      4.857       5191.0  no  
3          93.994          -36.4      4.857       5191.0  no  
4          93.994          -36.4      4.857       5191.0  no  

[5 rows x 21 columns]

In [29]:
sales.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 41188 entries, 0 to 41187
Data columns (total 21 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   age             41188 non-null  int64  
 1   job             41188 non-null  int32  
 2   marital         41188 non-null  int32  
 3   education       41188 non-null  int32  
 4   default         41188 non-null  int32  
 5   housing         41188 non-null  int32  
 6   loan            41188 non-null  int32  
 7   contact         41188 non-null  int32  
 8   month           41188 non-null  int32  
 9   day_of_week     41188 non-null  int32  
 10  duration        41188 non-null  int64  
 11  campaign        41188 non-null  int64  
 12  pdays           41188 non-null  int64  
 13  previous        41188 non-null  int64  
 14  poutcome        41188 non-null  int32  
 15  emp_var_rate    41188 non-null  float64
 16  cons_price_idx  41188 non-null  float64
 17  cons_conf_idx   41188 non-null 

In [30]:
y=sales.y
x=sales.drop("y",axis=1)

<h3>2) e. Standardize the data using the StandardScaler </h3>

In [60]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler().fit(x)
scaled_data=scaler.transform(x)
x= pd.DataFrame(scaled_data)

In [62]:
x.columns=['age','job','marital','education','default','housing','loan','contact','month','day_of_week','duration','campaign','pdays','previous','poutcome','emp_var_rate','cons_price_idx','cons_conf_idx','euribor3m','nr_employed'      ]

In [63]:
from sklearn.feature_selection import SelectFromModel
x_train,x_test,y_train,y_test= train_test_split(x,y,train_size = 0.8, test_size = 0.2,random_state =40)  

<h3> 2) c. Selecting important features based on Random Forest</h3>

In [64]:
sel = SelectFromModel(RandomForestClassifier(n_estimators = 100))
sel.fit(x_train, y_train)
sel.get_support()
selected_feat= x_train.columns[(sel.get_support())]
len(selected_feat)
selected_feat

Index(['age', 'duration', 'euribor3m', 'nr_employed'], dtype='object')

<h3>Q3. Build the following Supervised Learning models </h3>

<H3>a. Logistic Regression</H3>

In [53]:
#Linear regression
model=LogisticRegression().fit(x_train,y_train)
model.score(x_test,y_test)

0.9024034959941734

<H3>  b. AdaBoost </H3>

In [54]:
#AdaBoostRegressor
model1 = AdaBoostClassifier().fit(x_train,y_train)
model1.score(x_test,y_test)

0.9044671036659383

<H3> c. Naïve Bayes</H3>

In [55]:
 # Naïve Bayes
from sklearn.naive_bayes import GaussianNB
gnb = GaussianNB()
gnb.fit(x_train,y_train)
gnb.score(x_test,y_test)

0.8442583151250304

<H3>   d. SVM</H3>

In [57]:
#SVM
from sklearn import svm
clf = svm.SVC(kernel='linear')
clf.fit(x_train, y_train)
clf.score(x_test,y_test)

0.8923282350084972

<H2>Building the Supervised Learning models only with the selected features </H2>

In [65]:
x1=x[['age', 'duration', 'euribor3m', 'nr_employed']]

In [66]:
x1_train,x1_test,y_train,y_test= train_test_split(x1,y,train_size = 0.8, test_size = 0.2,random_state =40)  

<H3>a. Logistic Regression</H3>

In [67]:
#Linear regression
model=LogisticRegression().fit(x1_train,y_train)
model.score(x1_test,y_test)

0.8968196164117505

<H3>  b. AdaBoost </H3>

In [68]:
#AdaBoostRegressor
model1 = AdaBoostClassifier().fit(x1_train,y_train)
model1.score(x1_test,y_test)

0.8964554503520272

<H3> c. Naïve Bayes</H3>

In [69]:
 # Naïve Bayes
from sklearn.naive_bayes import GaussianNB
gnb = GaussianNB()
gnb.fit(x1_train,y_train)
gnb.score(x1_test,y_test)

0.8882010196649672

<H3>   d. SVM</H3>

In [70]:
#SVM
from sklearn import svm
clf = svm.SVC(kernel='linear')
clf.fit(x1_train, y_train)
clf.score(x1_test,y_test)

0.8861374119932023

<H2>Q4. </H2>  <H3> Here by comparing the scores we conclude that Adaboost is the best model with the score of 0.9044671036659383 </H3>